ibis/pyspark/compiler.py

```diff
@@ -0,0 +1,514 @@
+import collections
+import enum
+import functools
+
+import pyspark.sql.functions as F
+
+import ibis.common.exceptions as com
+import ibis.expr.operations as ops
+import ibis.expr.types as types
+from ibis.pyspark.operations import PySparkTable
+
+
+class AggregationContext(enum.Enum):
+    ENTIRE = 0
+    WINDOW = 1
+    GROUP = 2
+
+
+class PySparkExprTranslator:
+    _registry = {}
+
+    @classmethod
+    def compiles(cls, klass):
+        def decorator(f):
+            cls._registry[klass] = f
+            return f
+
+        return decorator
+
+    def translate(self, expr, scope, **kwargs):
+        # The operation node type the typed expression wraps
+        op = expr.op()
+
+        if type(op) in self._registry:
+            formatter = self._registry[type(op)]
+            return formatter(self, expr, scope, **kwargs)
+        else:
+            raise com.OperationNotDefinedError(
+                'No translation rule for {}'.format(type(op))
+            )
+
+
+compiles = PySparkExprTranslator.compiles
+
+
+def compile_with_scope(t, expr, scope):
+    """Compile a expression and put the result in scope.
+
+       If the expression is already in scope, return it.
+    """
+    op = expr.op()
+
+    if op in scope:
+        result = scope[op]
+    else:
+        result = t.translate(expr, scope)
+        scope[op] = result
+
+    return result
+
+
+@compiles(PySparkTable)
+def compile_datasource(t, expr, scope):
+    op = expr.op()
+    name, _, client = op.args
+    return client._session.table(name)
+
+
+@compiles(ops.Selection)
+def compile_selection(t, expr, scope, **kwargs):
+    # Cache compile results for tables
+    op = expr.op()
+
+    # TODO: Support predicates and sort_keys
+    if len(op.predicates) > 0 or len(op.sort_keys) > 0:
```

Comment:
this should be `if op.predicates or op.sort_keys`

Reply:
Fixed